doc_internal/fish-riir-plan.md

```diff
@@ -0,0 +1,77 @@
+These is a proposed port of fish-shell from C++ to Rust, and from CMake to cargo or related. This document is high level - see the Development Guide for more details.
+
+## Why Port
+
+- Gain access to more contributors and enable easier contributions. C++ is becoming a legacy language.
+- Free us from the annoyances of C++/CMake, and old toolchains.
+- Ensure fish continues to be perceived as modern and relevant.
+- Unlock concurrent mode (see below).
+
+## Why Rust
+
+- Rust is a systems programming language with broad platform support, a large community, and a relatively high probability of still being relevant in a decade.
+- Rust has a unique strength in its thread safety features, which is the missing piece to enable concurrent mode - see below.
```

Comment:
Please let me point you towards [this comment](https://github.com/fish-shell/fish-shell/pull/9512#issuecomment-1410820102), specifically:

> Other languages either have some specific pain point (go, python, java) or simply aren't known to us - this includes D, Zig, Nim and waaaayyy too many to list. That means asking us to use them means getting everyone to learn them and so far none of us have been interested.

and

> We're not making an Objective Review For All Uses For Everyone, we are choosing what works for us. And so, social questions matter a lot. On that note rust also has a bunch of buzz and mindshare that other languages lack - would we have had people queuing up to help if we had gone with D? I doubt it.

This plan is *our plan*, for *our* use, and we have decided on rust, because it fits and we already have people who know it and like it, unlike Zig.

The choice of implementation language is entirely up to the implementors.

That means further discussions of other languages are out of scope for this issue. I will be hiding this review and I would like for you to leave it at that.

Reply:
1. This is not about the choice, but being aware of the tradeoffs. Specifically knowing the footguns of your tools and being able to technically describe them.
2. Being dismissive on technical aspects of the critique will not improve things. Attention is a good tool to fix pain points of a language, which is in Rust unsafe handling of integer arithmetic.

Reply:
> Marking it as off-topic feels rude to me.

This is marked off-topic because it *is* - we are not looking for consultation on the choice of programming language, except from those who write fish (who are all happy with rust) and packagers (who so far also seem to be okay).

In particular we aren't looking to be convinced to try another language - we have people who already know rust and have already put in a considerable amount of work to evaluate it for fish, and we do not have any who know zig. We have already explained this in two separate places, one of which I explicitly pointed you towards.

It is also off-topic because you do not know the project and so you do not know that integer arithmetic has never been a particular pain point in practice *for us*, even in C++, so your technical commentary has very little relevance. Integer arithmetic simply isn't a huge deal in shell programming.

The reality of the situation is that you are not a stakeholder and so you don't get a say in our choice of programming language. We don't even have to convince you.
Personally, I would recommend that you take a step back and disengage because it doesn't matter to you.

We explicitly explained that this is about *our* particular choice of technology and not commentary on these as a whole. As far as I know zig is fine, I wish it success. Just not for fish.

This will be the last message on the matter. I trust you won't reply or I will have to make more drastic moderation decisions.